packages/eots/src/eots.rs

```diff
@@ -272,98 +335,164 @@ impl PublicKey {
         point_to_bytes(&self.inner.to_projective()).to_vec()
     }
 
-    /// Verifies whether the given signature w.r.t. the public key, public randomness and message hash.
-    pub fn verify(&self, pub_rand: &[u8], msg_hash: &[u8], sig: &[u8]) -> Result<bool> {
-        let msg_hash: [u8; 32] = msg_hash
-            .try_into()
-            .map_err(|_| Error::InvalidInputLength(msg_hash.len()))?;
+    /// Verifies that the signature is valid for this message, public key and random value.
+    /// Precondition: r must be normalized
+    pub fn verify(&self, r_bytes: &[u8], message: &[u8], sig: &[u8]) -> Result<bool> {
+        let h = hash(message);
+        self.verify_hash(r_bytes, h, sig)
+    }
+
+    /// Verifies that the signature is valid for this hashed message, public key and random value.
+    /// Based on the unexported schnorr verification function from btcd.
+    pub fn verify_hash(&self, r_bytes: &[u8], hash: [u8; 32], sig: &[u8]) -> Result<bool> {
+        // Step 2.
+        //
+        // P = lift_x(int(pk))
+        //
+        // Fail if P is not a point on the curve
+        // (This is handled by the k256 crate when creating the PublicKey)
+
+        // Fail if P is not a point on the curve
+        // (Already verified by construction)
+
+        // Parse public randomness
+        let r = PubRand::new(r_bytes)?;
+
+        // Parse signature
+        let s = Signature::new(sig)?;
+
+        // Fail if r >= p is already handled by the fact r is a field element.
+        // Fail if s >= n is already handled by the fact s is a mod n scalar.
+
+        // e = int(tagged_hash("BIP0340/challenge", bytes(r) || bytes(P) || M)) mod n.
+        let r_point_bytes = r.to_bytes();
         let p_bytes = self.to_bytes();
-        let r = PubRand::new(pub_rand)?;
-        let r_bytes = r.to_bytes();
-        let c = <Scalar as Reduce<U256>>::reduce_bytes(
+
+        let e = <Scalar as Reduce<U256>>::reduce_bytes(
             &tagged_hash(CHALLENGE_TAG)
-                .chain_update(r_bytes)
+                .chain_update(r_point_bytes)
                 .chain_update(p_bytes)
-                .chain_update(msg_hash)
+                .chain_update(hash)
                 .finalize(),
         );
 
-        let s = Signature::new(sig)?;
-        let recovered_r =
-            ProjectivePoint::mul_by_generator(&*s) - self.inner.to_projective().mul(c);
+        // Negate e here so we can use AddNonConst below to subtract the s*G
+        // point from e*P.
+        let e_neg = -e;
+
+        // R = s*G - e*P
+        let s_g = ProjectivePoint::mul_by_generator(&*s);
+        let e_p = self.inner.to_projective().mul(e_neg);
+        let recovered_r = s_g + e_p;
+
+        // Fail if R is the point at infinity
+        if recovered_r.is_identity().into() {
```

Comment:
is point at `is_identity`  equivalent to point at infinity ?

Reply:
Yes https://crypto.stackexchange.com/questions/66288/what-is-the-point-at-infinity-on-secp256k1-and-how-to-calculate-it